### PR TITLE
Mail assignee when issue/pull request is assigned

### LINKED ISSUE #8233: Send Email to Assignee on Assignment
## Description

This is a feature request to send an email to an assignee when that person is assigned an issue (unless they assigned themselves).

This is referenced in and similar to https://github.com/go-gitea/gitea/issues/5083 but I think it deserves it's own issue as it could be implemented independent of that more complex issue.

Based on our user's feedback the expected behavior is that assigning an issue to someone will email them.  Otherwise, you have to assign someone and then do something like write a comment to let them know you've assigned it to them (that's redundant and painful).

Seems like this would be straightforward to add into https://github.com/go-gitea/gitea/blob/master/modules/notification/mail/mail.go and would make the email notifications more intuitive.

### PR DESCRIPTION
**Pull request by davidsvantesson**
Fix #8233.

Sends mail to assignee (only) when issue/pr is assigned. 
Should this generate a notification in the UI as well? Currently there seem to be no UI notifications where only one user is notified.

I had to refactor more than I expected for this one, not sure if everything is in correct places.

### CODE REVIEW

File: routers/repo/issue.go [hunk 1]
```diff
@@ -1130,10 +1130,20 @@ func UpdateIssueAssignee(ctx *context.Context) {
...
 				ctx.ServerError("ChangeAssignee", err)
 				return
 			}
+
+			assignee, err := models.GetUserByID(assigneeID)
+			if err != nil {
+				ctx.ServerError("GetUserByID", err)
+				return
+			}
+			if setting.Service.EnableNotifyMail && !assignee.IsOrganization() && assignee.EmailNotifications() == models.EmailNotificationsEnabled {
```
Comment (lunny) [inline routers/repo/issue.go]: The mail service check should be in mailNotification init, but not here.

File: services/issue/issue.go [hunk 2]
```diff
@@ -96,3 +98,84 @@ func ChangeTitle(issue *models.Issue, doer *models.User, title string) (err erro
...
+		if err != nil {
+			return err
+		}
+
+		assignee, err := models.GetUserByID(assigneeID)
+		if err != nil {
+			return err
+		}
+
+		if setting.Service.EnableNotifyMail && !assignee.IsOrganization() && assignee.EmailNotifications() == models.EmailNotificationsEnabled {
```
Comment (lunny) [inline services/issue/issue.go]: As above.

File: modules/notification/notification.go [hunk 3]
```diff
@@ -26,7 +27,9 @@ func RegisterNotifier(notifier base.Notifier) {
 
 func init() {
 	RegisterNotifier(ui.NewNotifier())
-	RegisterNotifier(mail.NewNotifier())
+	if setting.Service.EnableNotifyMail {
+		RegisterNotifier(mail.NewNotifier())
+	}
```
Comment (davidsvantesson) [inline modules/notification/notification.go:33]: setting not initialized from app.ini in init function, so this doesn't work

Comment (guillep2k) [reply modules/notification/notification.go:33]: Perhaps this should be moved to an exported function that will be called from `routers.NewServices()` ?

Comment (davidsvantesson) [reply modules/notification/notification.go:33]: Thanks, that seem reasonable.

**Comment by guillep2k** [comment]
@davidsvantesson Please check [my comment here](https://github.com/go-gitea/gitea/pull/8329#issuecomment-543332960). I've cross-commented inadvertently.

**Comment by davidsvantesson** [comment]
> @davidsvantesson Please check [my comment here](https://github.com/go-gitea/gitea/pull/8329#issuecomment-543332960). I've cross-commented inadvertently.

It actually exists a comment for the assignment so I should get that one somehow and use.

File: modules/notification/notification.go [hunk 4]
```diff
@@ -26,7 +27,9 @@ func RegisterNotifier(notifier base.Notifier) {
 
 func init() {
 	RegisterNotifier(ui.NewNotifier())
-	RegisterNotifier(mail.NewNotifier())
+	if setting.Service.EnableNotifyMail {
```
Comment (lunny) [inline modules/notification/notification.go:31]: This must be changed.

**Review by lunny — CHANGES_REQUESTED** [review]

File: models/issue_assignees.go [hunk 5]
```diff
@@ -110,73 +110,63 @@ func clearAssigneeByUserID(sess *xorm.Session, userID int64) (err error) {
...
 func UpdateAssignee(issue *Issue, doer *User, assigneeID int64) (err error) {
-	return issue.ChangeAssignee(doer, assigneeID)
+	_, _, err = issue.ChangeAssignee(doer, assigneeID)
+	return err
 }
 
 // ChangeAssignee changes the Assignee of this issue.
-func (issue *Issue) ChangeAssignee(doer *User, assigneeID int64) (err error) {
+func (issue *Issue) ChangeAssignee(doer *User, assigneeID int64) (removed bool, comment *Comment, err error) {
 	sess := x.NewSession()
```
Comment (guillep2k) [inline models/issue_assignees.go:118]: This is an opinion of mine, and I can be convinced otherwise. I'm not sure about moving the assignee creation outside of the _transaction_ for issue creation . If we do it this way (which has some good points as well), and there is an error after committing the issue, there will be an 500 Server Error page returned to the user, but the issue will have been created. To address this I'd see to changing the way the page works by returning the created issue and an error flash instead of the 500 Server Error page (e.g. "unable to add assignees"). That would prevent the user from creating the issue a second time. Conceptually, it looks like material for another PR, but this PR is already changing the transaction order, so...

Otherwise we could be using the new `models.DBContext` type and move all this up the chaing, but the issue insert retry code must be moved up the chain as well.

File: models/issue_assignees.go [hunk 6]
```diff
@@ -110,73 +110,63 @@ func clearAssigneeByUserID(sess *xorm.Session, userID int64) (err error) {
...
-	return nil
+
+	return removed, comment, nil
 }
 
-func (issue *Issue) changeAssignee(sess *xorm.Session, doer *User, assigneeID int64, isCreate bool) (err error) {
+func (issue *Issue) changeAssignee(sess *xorm.Session, doer *User, assigneeID int64, isCreate bool) (removed bool, comment *Comment, err error) {
 	// Update the assignee
-	removed, err := updateIssueAssignee(sess, issue, assigneeID)
+	removed, err = updateIssueAssignee(sess, issue, assigneeID)
```
Comment (guillep2k) [inline models/issue_assignees.go]: I know this function name comes from old code, but it looks like it's more `toggleIssueAssignee()` than `updateIssueAssignee()`.

File: models/issue_assignees.go [hunk 7]
```diff
@@ -192,7 +182,7 @@ func (issue *Issue) changeAssignee(sess *xorm.Session, doer *User, assigneeID in
 		}
 		if err := prepareWebhooks(sess, issue.Repo, HookEventPullRequest, apiPullRequest); err != nil {
```
Comment (guillep2k) [inline models/issue_assignees.go:180]: The webhooks part doesn't seem related with changing assignees. I thihk it should be moved to another function.

File: routers/api/v1/repo/issue.go [hunk 8]
```diff
@@ -218,7 +218,7 @@ func CreateIssue(ctx *context.APIContext, form api.CreateIssueOption) {
 		form.Labels = make([]int64, 0)
 	}
 
-	if err := issue_service.NewIssue(ctx.Repo.Repository, issue, form.Labels, assigneeIDs, nil); err != nil {
+	if err := issue_service.NewIssue(ctx.Repo.Repository, issue, form.Labels, nil); err != nil {
```
Comment (guillep2k) [inline routers/api/v1/repo/issue.go:240]: Just a thought. I imagine that the assignee creation and webhook triggering calls could be moved inside `issue_service.NewIssue()`, unless there are other scenarios I fail to see.

File: routers/api/v1/repo/pull.go [hunk 9]
```diff
@@ -299,6 +300,8 @@ func CreatePullRequest(ctx *context.APIContext, form api.CreatePullRequestOption
 		return
 	}
 
+	issue_service.AddAssignees(prIssue, ctx.User, assigneeIDs)
```
Comment (guillep2k) [inline routers/api/v1/repo/pull.go]: Same here.

File: services/mailer/mail.go [hunk 10]
```diff
@@ -220,3 +222,8 @@ func SendIssueMentionMail(issue *models.Issue, doer *models.User, content string
 	}
 	SendAsync(composeIssueCommentMessage(issue, doer, content, comment, mailIssueMention, tos, "issue mention"))
 }
+
+// SendIssueAssignedMail composes and sends issue assigned email
+func SendIssueAssignedMail(issue *models.Issue, doer *models.User, content string, comment *models.Comment, tos []string) {
```
Comment (guillep2k) [inline services/mailer/mail.go:227]: Could we consider a "issue was unassigned" notification? It's a less common scenario, though, unless it's an operation error from the part of the doer.

**Comment by davidsvantesson** [comment]
Regarding assignees when issue is created:

- My current solution is to first add all assignees (which also trigger corresponding notification) and then do Notify New Issue. The assigned notification will come before the new issue notification. Gmail handles the threading well as the "new issue" mail has the Message-ID, other clients can get it in wrong order. Mail order is not guaranteed over network anyhow.
- Other options is to first send "New Issue Notifications" and then add assignees. People that are not participants by other means will only get the assigned mail.
- Of course other solutions is possible.

File: models/issue_assignees.go (hunk 6 above)
Comment (davidsvantesson) [reply models/issue_assignees.go]: Yes, I need to look over this functions.

File: models/issue_assignees.go (hunk 5 above)
Comment (davidsvantesson) [reply models/issue_assignees.go:118]: I need the "assigned comment" of each added assignee when doing the notification. So if moving this down to models again I need to move the notification part as well.

Also see my [comment below](https://github.com/go-gitea/gitea/pull/8546#issuecomment-543864362) regarding notification and order. 

DBContext means a transaction you do in whole or it will be rewinded?

File: services/mailer/mail.go (hunk 10 above)
Comment (davidsvantesson) [reply services/mailer/mail.go:227]: Do you really want that by default? There is no refined mail settings that can make that an option for the user. At least for me, if I am not assigned any longer I probably don't care :)

File: models/issue_assignees.go (hunk 5 above)
Comment (davidsvantesson) [reply models/issue_assignees.go:118]: A note is that issue events (like added assignees) does not have an id in the html ("issuecomment-X") so the href hash part doesn't work. But that could easily be added.

**Comment by guillep2k** [comment]
> 
> 
> Regarding assignees when issue is created:
> 
> * My current solution is to first add all assignees (which also trigger corresponding notification) and then do Notify New Issue. The assigned notification will come before the new issue notification. Gmail handles the threading well as the "new issue" mail has the Message-ID, other clients can get it in wrong order. Mail order is not guaranteed over network anyhow.
> 
> * Other options is to first send "New Issue Notifications" and then add assignees. People that are not participants by other means will only get the assigned mail.
> 
> * Of course other solutions is possible.

I was not referring to the e-mails, but the rows created in the database. Let me illustrate:

* User: opens "new issue", fills in the form, flags assignees, presses "CREATE".
* `routers.NewIssuePost()`: calls `issue_service.NewIssue()`.
* `models.NewIssue()`: opens a transaction, inserts the new issue, commits, closes transaction.
* `routers.NewIssuePost()`: calls `issue_service.AddAssignees()`; the transaction may fail for whatever reason.
* The user gets a `500 Server Error` page, _**but the Issue already exists!**_

Before this PR, the issue was created or not created, but if anything went wrong when assigning people, the whole issue was rolled back.

My comment was related to the code refactoring you did, not the notifications themselves. Sorry if I wasn't clear before.

**Comment by davidsvantesson** [comment]
I understood the issue, I just wanted to try give reason why I moved it up in the chain.
Surprisingly, if the assignee doesn't have right access the issue is never created, because that is already checked in "ValidateRepoMetas".
As I saw some discussion elsewhere, for "expected fails" like lacking access we should send back proper feedback to the user. There could still be unexpected failures (bugs) that can cause half-done transactions so it needs to be considered.

File: models/issue_assignees.go (hunk 5 above)
Comment (guillep2k) [reply models/issue_assignees.go:118]: I'm sorry, I've missed these comments. Sometimes I come in from "Notifications" and sometimes from my e-mail backlog... with different degrees of success. 😝 

I understand your problem. I have no objections about taking the add assignee part out of the issue.

File: routers/api/v1/repo/issue.go (hunk 8 above)
Comment (guillep2k) [inline routers/api/v1/repo/issue.go:240]: Assignements are handled _after_ the issue is committed, alright. But assignee validation should occur before that. Otherwise someone using the API will realize too late that they made a mistake and we don't yet have a "delete issue" option. If I understood your code correctly, you've removed that validation from `models.newIssue()` (which makes sense because it will no longer create assignees), but I think they must be validated beforehand nonetheless. This should apply to all the other calls to `models.NewIssue()`

File: services/issue/issue.go [hunk 11]
```diff
@@ -96,3 +97,109 @@ func ChangeTitle(issue *models.Issue, doer *models.User, title string) (err erro
...
+// AddAssigneeIfNotAssigned adds an assignee only if he isn't already assigned to the issue.
+// Also checks for access of assigned user
+func AddAssigneeIfNotAssigned(issue *models.Issue, doer *models.User, assigneeID int64) (err error) {
+	// Check if the user is already assigned
+	isAssigned, err := models.IsUserAssignedToIssue(issue, &models.User{ID: assigneeID})
+	if err != nil {
+		return err
+	}
+
+	if !isAssigned {
```
Comment (guillep2k) [inline services/issue/issue.go]:
```suggestion
	if isAssigned {
		return nil
	}
```
Is neater, I think

File: routers/api/v1/repo/issue.go (hunk 8 above)
Comment (guillep2k) [reply routers/api/v1/repo/issue.go:240]: > Surprisingly, if the assignee doesn't have right access the issue is never created, because that is already checked in "ValidateRepoMetas".

And NOW I've seen your comment. 😅

**Review by guillep2k — APPROVED** [review]

File: models/repo_permission.go [hunk 12]
```diff
@@ -329,10 +329,23 @@ func HasAccessUnit(user *User, repo *Repository, unitType UnitType, testMode Acc
 	return hasAccessUnit(x, user, repo, unitType, testMode)
 }
 
-// canBeAssigned return true if user could be assigned to a repo
+func canBeAssignedIssues(e Engine, user *User, repo *Repository) (bool, error) {
+	return hasAccessUnit(e, user, repo, UnitTypeIssues, AccessModeWrite)
```
Comment (lunny) [inline models/repo_permission.go]: I don't think this change is right. Who can be assigned for an issue? We should consider it carefully.

File: services/mailer/mail.go (hunk 10 above)
Comment (lunny) [reply services/mailer/mail.go:227]: Is the assign/unassign mail only sent to assignee? it Maybe also sent to others?

File: models/issue.go [hunk 13]
```diff
@@ -984,7 +984,6 @@ type NewIssueOptions struct {
 	Repo        *Repository
 	Issue       *Issue
 	LabelIDs    []int64
-	AssigneeIDs []int64
```
Comment (lunny) [inline models/issue.go:899]: Why removed this?

**Review by lunny — CHANGES_REQUESTED** [review]

File: modules/notification/notification.go (hunk 4 above)
Comment (davidsvantesson) [reply modules/notification/notification.go:31]: I call this initialization from routers/init.go now.

File: models/issue.go (hunk 13 above)
Comment (davidsvantesson) [reply models/issue.go:899]: I moved adding assignees outside the creation of new issues. Reason is that each added assignee creates a new comment (issue event) which is used in the notification.

File: models/repo_permission.go (hunk 12 above)
Comment (guillep2k) [reply models/repo_permission.go]: @lunny What are your thoughts? I didn't think about it before, but perhaps an Issue can be assigned to someone from QA or customer service that not necessarily will need to create PRs (e.g just gather information). They will probably not even need access to the code.
A more important consideration is who can _assign_ other people to an issue, as they need to have the authority to do so. I believe we're still not making that distinction .

Comment (davidsvantesson) [reply models/repo_permission.go]: Private repositories: At least read access to issues (or pull requests) should be needed so you can at least see the issue/pr. You can have access to code but not read access to issues. I don't know if it makes sense to be able to assign someone which does not have write permission in the issue/pr (can you still make comments?) 
Public repositories: Anyone has at least read access so the same applies.
For me it is OK to be less restrictive and allow assigning anyone with read access, but maybe it should be a separate PR. Should I change back to UnitTypeCode to keep existing behavior (just refactor now)?

Comment (davidsvantesson) [reply models/repo_permission.go]: The current code is very complex. It is already required that assignee has write access to issue / pr. It is checked by ValidateRepoMetas in routers/repo/issue.go. So I think this code shall be removed and use existing function instead.

Comment (guillep2k) [reply models/repo_permission.go]: Perhaps the criteria for "can be assigned" should be "write on this issue OR (read this issue AND create a PR)". Because a PR can close an issue by means of commit messages when it's merged; write access to the issue is not _really_ required. In any case, if you're assigned you should have some way of closing it or comment on it for others to close.
Just for the sake of discussion, anyone assigned should be allowed to comment on it, even if they wouldn't be able to comment on other issues (but that's not for this PR).

Comment (davidsvantesson) [reply models/repo_permission.go]: I think the check of "is assignable" should match users shown in the UI? The UI list is currently only based on any write access to the repository.

Comment (guillep2k) [reply models/repo_permission.go]: I agree they should match, either way we want that criteria to be.

File: models/issue_assignees.go (hunk 7 above)
Comment (davidsvantesson) [reply models/issue_assignees.go:180]: Actually it is, below is help text for PR trigger event. :wink:

> Pull request opened, closed, reopened, edited, approved, rejected, review comment, **assigned, unassigned**, label updated, label cleared or synchronized.

File: models/repo_permission.go (hunk 12 above)
Comment (davidsvantesson) [reply models/repo_permission.go]: I have relaxed the required for whom is assignable to anyone with write access to code, issues or pull requests. The UI only check for any write access to repo so this is to get as close as possible to that. (I know it is still not identical as I have not considered other units than these three). 
I think we shall leave additional changes of assignable out of this PR, and continue discussing this in #4329.

File: modules/notification/notification.go (hunk 4 above)
Comment (davidsvantesson) [reply modules/notification/notification.go:31]: New solution ok?

File: models/issue_assignees.go (hunk 6 above)
Comment (davidsvantesson) [reply models/issue_assignees.go]: I made additional refactoring here, I think it is clearer now what each function do.

File: services/mailer/mail.go (hunk 10 above)
Comment (davidsvantesson) [reply services/mailer/mail.go:227]: Issue was unassigned notification: I think it is a good idea but my opinion is it should be an option (in app.ini or elsewhere). It would be good if we can keep it out from this PR as it is already rather large.

Notification to others: I see this similar as the mention mail; being assigned is mostly interesting for the person being assigned. On the other hand update of assignees is a change of the issue, so it would make some sense to send notifications to all participants as well. 
The problem is that each added assignee is handled separate now, so if you add three new assignees it would trigger three notifications/mails. It would require a log of refactoring again to change that, and as I don't see it is very useful I would like to keep sending notification only to the assigned person.

File: models/issue_assignees.go [hunk 14]
```diff
@@ -58,7 +58,11 @@ func getAssigneesByIssue(e Engine, issue *Issue) (assignees []*User, err error)
 
 // IsUserAssignedToIssue returns true when the user is assigned to the issue
 func IsUserAssignedToIssue(issue *Issue, user *User) (isAssigned bool, err error) {
-	isAssigned, err = x.Exist(&IssueAssignees{IssueID: issue.ID, AssigneeID: user.ID})
+	return isUserAssignedToIssue(x, issue, user)
+}
+
+func isUserAssignedToIssue(e Engine, issue *Issue, user *User) (isAssigned bool, err error) {
+	isAssigned, err = e.Get(&IssueAssignees{IssueID: issue.ID, AssigneeID: user.ID})
```
Comment (guillep2k) [inline models/issue_assignees.go]:
```suggestion
	return e.Get(&IssueAssignees{IssueID: issue.ID, AssigneeID: user.ID})
```

File: models/issue_assignees.go (hunk 7 above)
Comment (guillep2k) [reply models/issue_assignees.go:180]: > 
> 
> Actually it is, below is help text for PR trigger event. 😉
> 
> > Pull request opened, closed, reopened, edited, approved, rejected, review comment, **assigned, unassigned**, label updated, label cleared or synchronized.

Oh, that's not what I meant. 😄 What I meant is that this code seems to belong to `services` rather than `models`. Anyway, it was here before, so refactoring should be for another PR.

File: models/issue_assignees.go [hunk 15]
```diff
@@ -208,67 +196,50 @@ func (issue *Issue) changeAssignee(sess *xorm.Session, doer *User, assigneeID in
...
+	if err != nil {
+		return false, err
 	}
 
-	// Loop through all assignees to add them
-	for _, assigneeName := range multipleAssignees {
-		assignee, err := GetUserByName(assigneeName)
-		if err != nil {
-			return err
+	// Check if the submitted user is already assigne, if yes delete him otherwise add him
```
Comment (guillep2k) [inline models/issue_assignees.go]:
```suggestion
	// Check if the submitted user is already assigned, if yes delete him otherwise add him
```

File: routers/api/v1/repo/issue.go [hunk 16]
```diff
@@ -336,9 +360,9 @@ func EditIssue(ctx *context.APIContext, form api.EditIssueOption) {
 			oneAssignee = *form.Assignee
 		}
 
-		err = models.UpdateAPIAssignee(issue, oneAssignee, form.Assignees, ctx.User)
+		err = issue_service.UpdateAssignees(issue, oneAssignee, form.Assignees, ctx.User)
```
Comment (guillep2k) [inline routers/api/v1/repo/issue.go:363]: Did I miss the `models.CanBeAssigned()` for this call?

File: routers/api/v1/repo/pull.go [hunk 17]
```diff
@@ -387,12 +411,12 @@ func EditPullRequest(ctx *context.APIContext, form api.EditPullRequestOption) {
 	// Send an empty array ([]) to clear all assignees from the Issue.
 
 	if ctx.Repo.CanWrite(models.UnitTypePullRequests) && (form.Assignees != nil || len(form.Assignee) > 0) {
-		err = models.UpdateAPIAssignee(issue, form.Assignee, form.Assignees, ctx.User)
+		err = issue_service.UpdateAssignees(issue, form.Assignee, form.Assignees, ctx.User)
```
Comment (guillep2k) [inline routers/api/v1/repo/pull.go:414]: Same here about calling `CanBeAssigned()`

File: templates/mail/issue/assigned.tmpl [hunk 18]
```diff
@@ -0,0 +1,17 @@
+<!DOCTYPE html>
+<html>
+<head>
+	<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
+	<title>{{.Subject}}</title>
+</head>
+
+<body>
+	<p>@{{.Doer.Name}} assigned you to the {{if eq .Issue.IsPull true}}pull request{{else}}issue{{end}}  <a href="{{.Link}}">#{{.Issue.Index}}</a>.</p>
```
Comment (guillep2k) [inline templates/mail/issue/assigned.tmpl]: I know it's in the subject, but could it be better to include the name of the repository here as well?

File: models/issue.go (hunk 13 above)
Comment (lunny) [reply models/issue.go:899]: I think adding assignee comment should be in the transaction but not in the notifications?

File: routers/api/v1/repo/issue.go (hunk 16 above)
Comment (davidsvantesson) [reply routers/api/v1/repo/issue.go:363]: For newly added assignees it is checked in UpdateAssignees->AddAssigneeIfNotAssigned. As this is not a new issue, there is no need to check before doing any transactions.

File: models/issue_assignees.go (hunk 7 above)
Comment (davidsvantesson) [reply models/issue_assignees.go:180]: Aha, yes you are probably right, but I will leave out that refactoring from this PR. I think we should have some design guidelines what type of code should be where. I think I've started to understand Gitea's code structure now....

File: models/issue_assignees.go (hunk 14 above)
Comment (davidsvantesson) [reply models/issue_assignees.go]: Fixed

File: models/issue_assignees.go (hunk 15 above)
Comment (davidsvantesson) [reply models/issue_assignees.go]: Fixed

File: templates/mail/issue/assigned.tmpl (hunk 18 above)
Comment (davidsvantesson) [reply templates/mail/issue/assigned.tmpl]: Added.

File: models/issue.go (hunk 13 above)
Comment (davidsvantesson) [reply models/issue.go:899]: I am not sure I understand now. The added assignee comment is used when sending the notification (the comment hash is included in the mail).

Comment (guillep2k) [reply models/issue.go:899]: ~@lunny If I've read that right, it's not a comment in the sense of a row in the `comment` table but a message for the contents of the e-mail. @davidsvantesson, perhaps it should be renamed to `message` to avoid confusion with the existing model/structure?~

Comment (davidsvantesson) [reply models/issue.go:899]: I am not sure if I still misunderstand you. Adding an assignee creates a Comment in the database of type CommentTypeAssignees. CommentTypeComment is another type. Maybe IssueEvent would have been a better name to reflect it is not only comments. Anyhow the Comment (of type Assignees) is used when constructing the e-mail.

Comment (guillep2k) [reply models/issue.go:899]: @davidsvantesson OK, sorry, I got confused about that. I understand @lunny's concern then. However, I'm not sure I'd block the creation of the issue because of an error while doing the assignments. That's why I think the error Flash is useful.

Comment (davidsvantesson) [reply models/issue.go:899]: The input data is validated before issue creation, but it is of course possible it can still fail because of unconsidered reasons / bugs. I think it is a good idea to use error flash.

Comment (davidsvantesson) [reply models/issue.go:899]: I've added an error flash now specifically for the case if add assignees fail when creating new issues/pr. Other error flashes I think can be for another PR. 
I am not sure how to handle it for API creation. Right now it can potentially give error although the issue itself is created.

**Comment by davidsvantesson** [comment]
I've done some more thorough testing of this now and generally I think it behaves as expected (also for error cases). 

One thing I noticed (not changed by this PR) is that assignees (users) that doesn't exist is ignored when creating an issue via the API, but it generates an error when updating the issue (assignees) via the API. What do you think about this behavior? At least I think it should be the same in both cases?

**Comment by guillep2k** [comment]
> 
> 
> I've done some more thorough testing of this now and generally I think it behaves as expected (also for error cases).

Great!

> 
> One thing I noticed (not changed by this PR) is that assignees (users) that doesn't exist is ignored when creating an issue via the API, but it generates an error when updating the issue (assignees) via the API. What do you think about this behavior? At least I think it should be the same in both cases?

I think that it should be consistent. They both should fail, but that's considered a breaking change so we should be careful about it.

**Comment by davidsvantesson** [comment]
I don't see why someone would have a use-case expecting the API to ignore non-existing users (as non-access users still generate error), so I don't think it has to be seen as breaking change.

**Comment by davidsvantesson** [comment]
> assignees (users) that doesn't exist is ignored when creating an issue via the API

I have fixed this now so it generates an error.

**Comment by lunny** [comment]
I don't know why I cannot received email when assigned a user but I can receive when click test mail from admin panel.

**Comment by davidsvantesson** [comment]
@lunny You made a similar comment in #8329. Can you confirm you have the same problem in both PRs?

**Comment by guillep2k** [comment]
@lunny I think there was a problem with one of the migrations regarding mail preferences. After one update I stopped receiving mails from gitea.com (and I'm watching all repos). I had to manually go to my profile and confirm my e-mail preference, even when it already displayed "Enable Email Notifications". I'm not the only one that had too, even (I don't remember who else ATM). Please check your database for inconsistencies regarding this column.

**Review by lunny — APPROVED** [review]
